pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java

```diff
@@ -0,0 +1,357 @@
+        List<Instrument> newInstruments = lims.getInstruments();
+        List<InstrumentModel> newInstrumentModels = lims.getInstrumentModels();
+        List<Box> newBoxes = lims.getBoxes();
+
+        List<SampleProvenance> newSampleProvenance =
+            ProvenanceUtils.buildSampleProvenance(
+                newProjects, newInstruments, newInstrumentModels, newOrders, newSamples, newRuns);
+        List<LaneProvenance> newLaneProvenance =
+            ProvenanceUtils.buildLaneProvenance(newInstruments, newInstrumentModels, newRuns);
+
+        synchronized (this) {
+          this.samples = newSamples;
+          this.projects = newProjects;
+          this.users = newUsers;
+          this.orders = newOrders;
+          this.runs = newRuns;
+          this.sampleTypes = newSampleTypes;
+          this.sampleAttributes = newSampleAttributes;
+          this.changeLogs = newChangeLogs;
+          this.instruments = newInstruments;
+          this.instrumentModels = newInstrumentModels;
+          this.boxes = newBoxes;
+          this.sampleProvenance = newSampleProvenance;
+          this.laneProvenance = newLaneProvenance;
+
+          cacheComplete = true;
+          lastUpdated = Instant.now();
+          cacheUpdateFailures.set(0);
+          log.debug("Update successful");
+        }
+      } catch (RuntimeException e) {
+        cacheUpdateFailures.inc();
+        throw e;
+      } finally {
+        synchronized (this) {
+          cacheUpdateTimer.observeDuration();
+          log.debug("Update attempt completed; notifying others");
+          updating = false;
+          notifyAll();
+        }
+      }
+    }
+  }
+
+  public synchronized void updateIfEmpty() {
+    if (!enabled) {
+      throw new IllegalStateException("Cannot retrieve data from disabled cache");
+    }
+    if (!cacheComplete) {
+      update();
+    }
+  }
+
+  @Override
+  public synchronized Sample getSample(String id) {
+    updateIfEmpty();
+    return getBy(samples, Sample::getId, id);
+  }
+
+  @Override
+  public synchronized List<SampleProject> getSampleProjects() {
+    updateIfEmpty();
+    return projects;
+  }
+
+  @Override
+  public synchronized List<Sample> getSamples(
+      Boolean archived,
+      Set<String> projects,
+      Set<String> types,
+      ZonedDateTime before,
+      ZonedDateTime after) {
+    updateIfEmpty();
+
+    Set<String> normalizedProjects = normalizeSet(projects);
+    Set<String> normalizedTypes = normalizeSet(types);
+    Date beforeDate = before == null ? null : Date.from(before.toInstant());
+    Date afterDate = after == null ? null : Date.from(after.toInstant());
+
+    return samples.stream()
+        .filter(
+            sample -> {
+              if (archived != null && !archived.equals(sample.getArchived())) {
+                return false;
+              }
+              if (normalizedProjects != null
+                  && normalizedProjects.stream()
```

Comment:
I think these can be reduced to a `!normalizedX.contains(sample.getX())`

Reply:
changed